### DISCUSSION TITLE
Specify sort by field on one to many field

### DISCUSSION THREAD

**Discussion by alexvdvalk** [directus/directus#22799] [Feature Requests]
### Summary

When adding a one to many field, I want to specify a field on which i can sort within the interface:

![Screenshot 2024-06-19 at 15 22 29](https://github.com/directus/directus/assets/21194068/0dcc79eb-1bb1-4827-882b-fc78c7401d8a)

This can be manually done in the table layout, but selected sort is not persisted when i navigate between items.

### Basic Example

When adding a One to many field, I can either specify and create a "SORT" field, which means users can manually sort the related items, or I can specify another field and direction which will be included in the API request when loading the one 2 many interface.

### Motivation

In cases where there are hundreds of related items in a one to many field, users see the labels as configured in the interface config, but the order seems arbitary.

### Detailed Design

Two additional fields on the field config:
 - option to select a field from the related collection
 - sort direction - Ascending or Descending

There will need to be a check in place that will ensure the manual sort is not used.

When the many to one interface loads, it will include the ?sort query parameter.

### Must Have:
- Ability to specify sort field and directus

### Should Have:
- Validation where users can select EITHER a manual 'sort' field or a 'sort by related value' field 

### Could Have:
- Ability to select multiple sort fields, which is already supported in the API already

### Won't Have:
- customisation on a user by user basis

### Drawbacks

None.

### Alternatives

The table layout allows users to sort by one of the available fields, but this is not persisted.

### Adoption Strategy

This shouldn't have any breaking changes.

### Unresolved Questions

What happens if a one 2 many already has a manual sort field setup?

**Comment by MarkBurvs** [comment]
This would be amazing. 
We rely on many instances of table view o2m fields in our project, and it's very annoying that the selected sort does not persist when you leave and return to the item.

**Comment by JSDA123** [comment]
We also rely on o2m tables views and it's rather painful not being able to create a predictable user experience in these cases.

Maybe this could be implemented on the Display configuration, specifying default sort and a column widths. In our case that would be enough. We don't necessarily need persistence of the kind available in the Bookmarks system for Collections pages.

**Reply by MarkBurvs** [reply]
Happily, it does look like this issue is being addressed @JSDA123 : https://github.com/directus/directus/pull/23019#issuecomment-2283266011

**Comment by carlo-salinari** [comment]
Joining the discussion late here.

I second @JSDA123's opinion: this would be much more intuitively handled in the Display configuration.

We have a relatively simple products catalogue database, with o2m, m2o, and m2m relationships.
Our clients are upset due to the confusing sorting behavior, and it's hitting us hard.

**Reply by MarkBurvs** [reply]
I think this has been fixed now in the latest version.

**Reply by carlo-salinari** [reply]
> I think this has been fixed now in the latest version.

I believe the fix only covers the data editing Interface for o2m (I'm running v11.1.0).
The problem is present in simple data Display as well, though.

If e.g. you have a Countries collection with ISO codes as PK:

| ISO (PK) | Name |
|----|-------|
| GB | United Kingdom|
| IT | Italy |
|---|---|

And a products Availability table that links countries to each product (m2o), sorting by name ascending in the Data App UI will show:

| Country ID > Name ↑ | Product |Available
|----|-------|----|
|  United Kingdom| product A | true|
| Italy | product A | false|
|---|---|---|

With "United Kingdom" coming before "Italy" (it picks the relative order of the primary keys GB < IT, even if they are not displayed) which is very puzzling for users.

**Reply by alexvdvalk** [reply]
You can edit the raw value of the sort and use the dot notation to sort by Country.Name. I've created a slightly different approach where you can do a multi level sort on the interface: https://github.com/directus/directus/pull/23957

**Reply by carlo-salinari** [reply]
> You can edit the raw value of the sort and use the dot notation to sort by Country.Name. I've created a slightly different approach where you can do a multi level sort on the interface: #23957

Thanks for the suggestion!
I’m not entirely clear on how this applies, though.

In my collection `availability_by_country`, the `country_id` field is a many-to-one (m2o) relationship, meaning that multiple availability records are linked to single countries and products.

![image](https://github.com/user-attachments/assets/16a45a9f-ae01-42a5-8488-477c64029eed)

![image](https://github.com/user-attachments/assets/52e4423b-55e1-4baf-a457-1cb4607f9927)

However, under **Interface** for the `country_id` field, there’s no Sort option to edit:

![image](https://github.com/user-attachments/assets/a8708c4c-db08-48e2-82a5-699214326b9e)

To clarify, I’ve recreated the issue below:

![image](https://github.com/user-attachments/assets/70676db2-9f6d-4e77-814b-cf1727d18bb5)

Here, `country_id.name` seems sorted in ascending order, but the actual sort is applied to the primary key (the ISO code) regardless. This feels like a bug—one that's particularly frustrating.

The correct order should be of course:
1. French Polynesia
2. Italy
3. United Kingdom

[running v11.1.2]

**Reply by alexvdvalk** [reply]
This is because you're using a display label to display the country name when you select the country_id field. You can traverse the relational and add in the country > name field into your layout, and then the sorting will work as expected.

**Reply by alexvdvalk** [reply]
This discussion is really about the One to Many interface where you're busy editing a record and you see a list / table of the related items.
